java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -126,141 +129,108 @@ public String toString() {
-    if (transitiveReduction.isEmpty()) {
-      // If no new combination, try with the symmetric (because transitive reduction only checks the operand on the left side.
-      transitiveReduction = symmetric().transitiveReduction(knownRelations, usedRelations);
-    }
-    return resolveState(transitiveReduction, usedRelations);
-  }
-
-  private RelationState relationStateForSameOperand() {
-    switch (kind) {
-      case EQUAL:
-      case GREATER_THAN_OR_EQUAL:
-      case LESS_THAN_OR_EQUAL:
-      case METHOD_EQUALS:
-        return RelationState.FULFILLED;
-      case NOT_EQUAL:
-      case GREATER_THAN:
-      case LESS_THAN:
-      case NOT_METHOD_EQUALS:
-        return RelationState.UNFULFILLED;
-      default:
-        throw new IllegalStateException("Binary relation kind unsupported" + kind);
-    }
-  }
-
-  /**
-   * Returns a new list of relations, built by combining one of the known relations that can be combined transitively with the receiver
-   * 
-   * @param knownRelations the list of relations known so far
-   * @param usedRelations the list of relation that have been used in past recursive checks
-   * @return a list of relations containing at least one new relation; empty if no new combination was found.
-   */
-  private Collection<BinaryRelation> transitiveReduction(Collection<BinaryRelation> knownRelations, Set<BinaryRelation> usedRelations) {
-    boolean changed = false;
-    List<BinaryRelation> result = new ArrayList<>();
-    for (BinaryRelation relation : knownRelations) {
-      boolean used = false;
-      if (leftOp.equals(relation.leftOp)) {
-        used = result.addAll(relation.combinedRelations(knownRelations, usedRelations));
-      } else if (leftOp.equals(relation.rightOp)) {
-        used = result.addAll(relation.symmetric().combinedRelations(knownRelations, usedRelations));
-      }
-      if (used) {
-        changed = true;
-      } else {
-        result.add(relation);
+      iterations++;
+      for (BinaryRelation relation : relations) {
+        RelationState result = relation.implies(this);
+        if (result.isDetermined()) {
+          return result;
+        }
       }
+      usedRelations.addAll(relations);
+      relations = deduceTransitiveRelations(usedRelations);
     }
-    return changed ? result : Collections.<BinaryRelation>emptyList();
+    return RelationState.UNDETERMINED;
   }
 
-  private Collection<BinaryRelation> combinedRelations(Collection<BinaryRelation> relations, Set<BinaryRelation> usedRelations) {
-    List<BinaryRelation> result = new ArrayList<>();
-    for (BinaryRelation relation : relations) {
-      if (!this.equals(relation)) {
-        BinaryRelation combined = combineUnordered(relation);
-        if (combined != null && !usedRelations.contains(combined)) {
-          result.add(combined);
+  static Set<BinaryRelation> deduceTransitiveRelations(Set<BinaryRelation> relations) {
+    Set<BinaryRelation> result = new HashSet<>();
```

Comment:
As discussed, this implementation is not really efficient and can be simplified to avoid two loops over the relations and generate AND test within the same test loop.